pretyping/globEnv.mli

```diff
@@ -0,0 +1,83 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Names
+open Environ
+open Evd
+open EConstr
+open Ltac_pretype
+
+val empty_lvar : ltac_var_map
+
+(** To embed constr in glob_constr *)
+
+val register_constr_interp0 :
+  ('r, 'g, 't) Genarg.genarg_type ->
+    (unbound_ltac_var_map -> env -> evar_map -> types -> 'g -> constr * evar_map) -> unit
+
```

Comment:
Why did you move the registering function to `GlobEnv`? It belongs semantically to pretyping.

Reply:
I found it nice to have `pretyping` not needing anything specific to `ltac`. So I wanted to move the interpretation function for tactics-in-term (exactly for genargs-in-terms) to `globEnv`. But then, I was needing the `constr_interp` registration known from `GlobEnv`.

At least two possible alternatives:
- to keep ltac in `pretyping` and to provide a `ltac : GlobEnv.t -> ltac_var_map` function for use in `pretyping.ml`
- to move registration before `GlobEnv` in a file which reflects that it is about typing (meaning also moving `ltac_var_map`)

Reply:
This function is not specific to Ltac, as it is also used by e.g. Ltac2 and (if memory serves) Mtac. The fact that it takes a `GlobEnv` is a deficiency of the API, nothing more, and should be handled using a specific instance of pretyping once we have a generic pretyper.

Reply:
> My understanding is that, thanks to open recursion, you'd like parts 2 and 3 to be eventually only in plugins (thus getting also rid of registering).

Indeed.

> So, are you saying that you'd prefer parts 2. and 3. of GlobEnv to remain in pretyping.ml (even if it means keeping ltacX-related code in pretyping) and GlobEnv deal only with the renaming for Evars?

Rather, I think that when we have open recursion we merge back the evar-renaming mechanism into pretyping and we move GlobEnv to the Ltac plugin.

> Regarding Ltac2, I was actually surprised that the passing of a renaming substitution through the ist in #7300 did not work out of the box also for Ltac2. Would you like me to investigate this or do you prefer Ltac2 to support a first-order (non-binding) view at variables?

It's a design choice of Ltac2 that term variables are not handled transparently. It has a dubious semantics and a terrible efficiency cost even when not used. I had to hack around the Ltac hardcoded implementation to make it work properly, actually.